### PR TITLE
Release global lock before throwing exceptions from the Unix WaitSystem

### LINKED ISSUE #70010: [NativeAOT] Enable Reflection library tests in Linux-ARM64
This test hang in CI Linux-arm64 container without much debugging info and currently its [disabled ](https://github.com/dotnet/runtime/blob/main/src/libraries/tests.proj#L395-L396)in our CI runs.

### PR DESCRIPTION
**Pull request by jkotas**
The exception throwing and handling can enter wait system recursively that will lead to deadlock.

Fixes #70010

### CODE REVIEW

**Comment by jkotas** [comment]
cc @dotnet/ilc-contrib

File: src/libraries/System.Private.CoreLib/src/System/Threading/WaitSubsystem.WaitableObject.Unix.cs [hunk 1]
```diff
@@ -506,27 +504,22 @@ public int Wait_Locked(ThreadWaitInfo waitInfo, int timeoutMilliseconds, bool in
...
 
                     waitableObjects = null; // no need to clear this anymore, RegisterWait / Wait will take over from here
+
                     waitInfo.RegisterWait(count, prioritize, waitForAll);
-                    needToWait = true;
+                    return waitInfo.Wait(timeoutMilliseconds, interruptible, isSleep: false, ref lockHolder);
                 }
                 finally
                 {
+                    lockHolder.Dispose();
```
Comment (VSadov) [inline src/libraries/System.Private.CoreLib/src/System/Threading/WaitSubsystem.WaitableObject.Unix.cs:513]: Should this be after clearing waitableObjects?

**Review by VSadov — APPROVED** [review]
LGTM. Thanks!

File: src/libraries/System.Private.CoreLib/src/System/Threading/WaitSubsystem.WaitableObject.Unix.cs (hunk 1 above)
Comment (jkotas) [reply src/libraries/System.Private.CoreLib/src/System/Threading/WaitSubsystem.WaitableObject.Unix.cs:513]: `waitableObjects` is thread local pooled array. It should be safe to clean it outside the lock.

**Pull request merged**